### DISCUSSION TITLE
Maybe vs strictNullChecks

### DISCUSSION THREAD

**Discussion by benwaffle** [true-myth/true-myth#260] [General]
Is there a good case for using the `Maybe` type over enabling strictNullChecks in typescript? This flag will require you to handle null and undefined as separate types from the type of your value.

**Comment by chriskrycho** [comment]
This is a great question! I don’t think that `strictNullChecks` itself particularly moves the needle, and indeed we designed True Myth *assuming* `strictNullChecks: true`. Working with `null` and `undefined` still has all the kinds of pain points we call out in the docs: the necessity of scattering checks throughout the entire codebase in particular, and the necessity of every operation with interacts with the data having to therefore be aware of and handle `null` and `undefined`. Being able to define functions which *don’t* have any knowledge of the “optionality” of the data and apply them via `map` etc. is still very powerful and useful.

What *does* make for an interesting question and analysis, though, is the addition of the [optional chaining `?.`][oc], [nullish coalescing `??`][nc], and [nullish assignment `??=`][na] operators. Those complicate the value proposition for `Maybe` a fair bit, because many times they provide briefer ways of doing the same things which `Maybe` does, and they can be (and indeed *are*) optimized directly by browsers, and they have built-in syntactical support in the language (both JS and TS!). That doesn’t change the fundamental tradeoff—you could more or less do all of the same things with the ternary operator before the introduction of those new operators!—but it *does* change the verbosity part of the tradeoff quite a bit.

For my own part, I still like `Maybe` very much because it does provide a lot of utilities for just working naturally with functions which don’t have to be aware of the optionality of the data, especially because in the absence of a pipe operator (someday, maybe!) you end up having to revert to the ternary operator: the new operators are all biased in a way that makes them more or less not usable with standalone functions. For a trivial example of what I mean by this:

```ts
import { mapOr, just } from 'true-myth/maybe';

const add1 = (a: number) => a + 1;

function withMaybe(maybe: Maybe<number>): number {
  return maybe.mapOr(0, add1);
}

function withOptional(optional: number | null): number {
  return optional ? add1(optional) : 0;
}
```

This is a toy example, of course, but it shows how the new operators don't help you *at all* with standalone functions. Notice that there’s no good way to use optional chaining or nullish coalescing here! The operators are biased toward optional *object types with methods*, and really hurt when you're working with primitives *or* want to apply functions to objects as arguments (rather than just methods). And that’s totally fine; the point is simply that `Maybe` does still provide benefits in a lot of cases.

That said, in the cases where optional chaining or nullish coalescing are applicable, they’re *much* nicer:

```ts
function withMaybe(maybe: Maybe<string>) {
  return maybe.mapOr(0, (s) => s.length));
}

function withOptional(optional: string | null) {
  return optional?.length ?? 0;
}
```

So: it depends on what you’re doing. Both work. `Maybe` definitely still provides a lot of value. But it may not be *as* useful as `Result` in a new codebase, and certainly has a higher friction-to-value-delivered ratio than it did before the introduction of the new operators.

[oc]: http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining
[nc]: http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator
[na]: http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_nullish_assignment
